### ISSUE TITLE
Very tricky player freeze

### ISSUE THREAD

**Issue by greenozon** [mrfearless/MediaPlayer#9]
I played with this case a bit to lower the number of steps, here is the summary

open up the player
it auto plays the video
press FF (10 sec) button until it comes to half or far of the movie duration
press couple of times (eg 3) the faster play speed button
press couple of times the frame step button
!!! alarma  smth strange happens here !!!
the player freezes in such a condition that you can't play it again
even more - you drag-drop or open new (or same video) - nothing happens!
freeze dream continues!

only restart helps

hope you'll be able to reproduce it (I used a short 1:35 vid for this test case)

**Comment by mrfearless** [comment]
I think I will change to allow step 10 when playing at rate x1 or lower to see if that helps. 

There is a possibility that due to conversion from nanoseconds to milliseconds for duration and position, that using step 10 could still exceed the duration and cause the same issue. But I think limiting it to just 1x play makes sense as the seekbar is there if someone wants to adjust position manually anyhow when playing at a faster rate.

Also I think limiting the max play rate to 4x also make sense as it will reduce choppiness of the video stream.

**Comment by greenozon** [comment]
so were you able to reproduce this tricky edge case?

Also is it possible to print somewhere the current play rate speed? would be nice to see it somehow

**Comment by mrfearless** [comment]
The rate speed is now shown to the right of the current time position in square brackets. I have adjusted the rate to min x0.125 and max to x4, as less than 0.125 or 0 in some cases caused a freeze/crash. The step 10 now only works if play speed is normal or below it.

**Comment by greenozon** [comment]
question:  what is the step when one click the rate up/down buttons?
in my case I see only those:
0.5
1 (not printed)
2
4

is it possible to make it more fine-grained?
eg step 0.5 or even 0.25 please

**Comment by mrfearless** [comment]
x1 is normal speed, hence why i didn't print it out.

The rate is doubled or halved when clicking the rate up/down buttons. So x2 and x4 for faster, x0.5, x0.25 and x0.125 for slower. The min and max rate are determined by the media engine and are reported when querying the media engine with MFPMediaPlayer_GetSupportedRates (a wrapper for the IMFPMediaPlayer::GetSupportedRates method). 

I'm not sure how it determines the min and max rates. But if the media engine supports as low as x0.125 then it will go down to that at the lowest rate. The max rate I encountered was x8, but the video and audio playback is just too choppy to be useful so I limited it to x4 at most.

On my Win10 Pro PC, the min rate for me is x0.125, so I can drop it down to that.

**Comment by greenozon** [comment]
Oh wow, I see...
I was not aware that step is up to the media engine and not the player itself..
Thanks for the insights

I'll play around with diff videos to see different min/max values

**Comment by greenozon** [comment]
OK, so in my case there are very limited speed values:
0.5
1
2
4

but, when I tried to play the .wmv  I was able to move far then 4 (despite there was no number like it is 1x but it was much much faster then 1)

**Comment by greenozon** [comment]
Another speed rate issue:
when having [4] and press the slower button for some reason it get dropped to x0.5 (instead of 2x)

**Comment by greenozon** [comment]
same issue 2x -> 0..5x (instead of 1x)

**Comment by greenozon** [comment]
IMPORTANT UPDATE:  the bug with speed rate IS NOT reproduced in x86 player! (above issues were seen in x64 one)

**Comment by mrfearless** [comment]
Ok thanks for letting me know, i'll look into it.

**Comment by mrfearless** [comment]
I have changed how this is implemented in MediaPlayer. The supported rates reported for a media item and the engine are now shown in a submenu on the main menu, the context menu and a dropdown menu from a playback speed button, instead of the faster/slower buttons, which have now been removed. 

This playback speed menu is dynamically generated when there is a media item loaded, otherwise it will show a "Unvailable" menu item - in whichever language is used. The play rates are limited to min of x0.125 and max of x4. The playback rates (if supported) will be shown as 0.125, 0.25, 0.5, 0.75, 1.0, 1.25, 1.5, 1.75, 2.0, 3.0, 4.0 in the menu.

This will be in the next release.
